### DISCUSSION TITLE
BGU - Multi Model Implementation

### DISCUSSION THREAD

**Discussion by amielsaa** [useocl/use#65] [Ideas]
Hello Lars,
We have started implementing the multi model feature, by adding the rule discussed with Mira in USEBase.gpart.
We added a few classes that correspond to the parsing rules, and we want to make sure we're on the right track.
I've attached the commits in our fork, thanks a lot!
The files we modified are: `USEBase.gpart, Context.java, ASTMultiModel.java, MMultiModel.java, ModelFactory.java`
commit : https://github.com/Gil4390/useBGU/commit/a76efda7d9a12e8fd86c2765c0e3c131fbb5af16

**Comment by h-man2** [comment]
Hello!

I reviewed the commit and it looks good to me.

Best regards!

Lars

**Comment by amielsaa** [comment]
Hello @h-man2,

We encountered some issues while working on the multi-model feature, and we'll be happy if you can give us your opinion.
**Issue 1:**
we found a bug in our implementation where an error is thrown if you have 2 internal models both with a class named 'A'.
we have a couple of solutions in mind and would like your opinion

- Option 1: change the way classes get their names - so that a class with the name 'A' in internal-model1 will be renamed to 'A_internal-model1' and so all the class's names will be unique (because the internal models' names are unique)

- Option 2: create a new Context object for each generated model - so our first model will be associated with 'ctx1' object, and our second model will be associated with a 'ctx2' object,  but with the same error and modelFactory pointers.

**Issue 2:**
The way Multi-Models should be loaded into USE, what's a better approach: 

- Option 1: To enable USE to open multiple regular model files and parse them as a multi-model (which means changing the way you implemented the file explorer).

- Option 2: make it so you can only open a single file with a multi-model inside, and it will be up to the user to understand the new syntax of the multi-model.  (if necessary, we can make an external tool to combine the multiple regular models into a single multi-model file)

If you have a better idea for each of the issues we'll be happy to hear them!

Thanks a lot for your help!

Amiel & Gil
